### DISCUSSION TITLE
Read error when running "Real Time Database" tutorials

### DISCUSSION THREAD

**Discussion by mjolnirPH** [SchneiderInfosystems/FB4D#99] [Q&A] [answered]
Hello!

I just recently "returned" to FB4D after about more than a year because I couldn't figure it out the last time. I'm happy and thankful that now there are video tutorials I can watch and finally get to use Firebase with Delphi.

Following the real time database tutorials and running the app, the Form caption is changed to "Listener was stopped - restart the application." I tried restarting the app several times, but it ends up the same. However, when I enter some text into edtDBMessage and click on btnWrite, the Firebase DB gets updated, and unlike what's in the demo, I never get a successful read.

I'm on a Windows 10 machine, recently freshly formatted and I'm using "vanilla" 10.4.2, with no additional components, yet 😄.

What do I need to do to fix this? 

Thanks for your help.

**Comment by SchneiderInfosystems** [comment, marked as answer]
Hi, For me it is unclear which sample application you try? 
RTDB_SimpleReadWrite? 
Can you see an error within the DBError event before receiving the stop event?
Do you have installed the Database Rules as explained in the wiki at https://github.com/SchneiderInfosystems/FB4D/wiki/RT-DB-Sample-Applications#1st-example-the-global-string

**Reply by mjolnirPH** [reply]
Hi. I was following the video "FB4D: First Real Time Database Demo Application." I started from scratch, as I wasn't aware that the source code for the RTDB_SimpleReadWrite application was available.

The DBError event doesn't seem to fire as the lblStatus.Text does not change to "Error: " + the error message.

As to the database rules, I have already changed both the read and write rules to true.

I will try the RTDB_SimpleReadWrite and see if there is any change.

UPDATE:

I tried the RTDB_SimpleReadWrite application, and here is the screenshot on first run:

![image](https://user-images.githubusercontent.com/64586768/119253924-fbb85a80-bbe5-11eb-86b9-c958e46080e0.png)

When I tried entering some new text and clicked on btnWrite, I got this:

![image](https://user-images.githubusercontent.com/64586768/119253952-2c988f80-bbe6-11eb-9198-a2956d1e12ec.png)

and verifying it with the RTDB:

![image](https://user-images.githubusercontent.com/64586768/119253970-489c3100-bbe6-11eb-9afa-5564addcf53b.png)

And running the application again (I just added a time stamp to the "Firebase RT DB connected" line)

![image](https://user-images.githubusercontent.com/64586768/119254054-c102f200-bbe6-11eb-98d4-3f707e6c7e97.png)

From what I gather, the DBError is fired by the TRTDBLIstenerThread.

Thanks for your reply.

**Reply by SchneiderInfosystems** [reply]
Thank you for this report! 

In fact, since Delphi 10.4.2 there was a new problem due to changes in ARC memory management. This leads to a too early release of IFirebaseEvent returned by IRealTimeDB.ListenForValueEvents. The solution was simple, introducing a new member variable to extend the lifetime of this object. Can you check the latest version?

**Reply by mjolnirPH** [reply]
Hi Christoph!

The fix posted worked for both apps.

Thanks a lot!
